dlib/dnn/layers.h

```diff
@@ -3331,24 +3331,19 @@ namespace dlib
         {
         }
 
-        template <typename SUBNET>
-        void forward(
-            const SUBNET& sub,
-            resizable_tensor& data_output
-        )
+        void forward_inplace(const tensor& input, tensor& output)
         {
-            data_output.copy_size(sub.get_output());
-            tt::mish(data_output, sub.get_output());
+            tt::mish(output, input);
         }
 
-        template <typename SUBNET>
-        void backward(
+        void backward_inplace(
+            const tensor& computed_output,
             const tensor& gradient_input,
-            SUBNET& sub,
+            tensor& data_grad,
             tensor&
         )
         {
-            tt::mish_gradient(sub.get_gradient_input(), sub.get_output(), gradient_input);
+            tt::mish_gradient(data_grad, computed_output, gradient_input);
```

Comment:
`tt::mish_gradient(data_grad, computed_output, gradient_input);`

isn't equivalent to `tt::mish_gradient(sub.get_gradient_input(), sub.get_output(), gradient_input);` since `computed_output` isn't `sub.get_output()`.  `computed_output` is the output from `tt::mish()` in this case.   You would need a different gradient function to compute it in-place.

Reply:
Oh, I thought this was handled here:
https://github.com/davisking/dlib/blob/3162f93c5d303e593337ab8e40d79b385e025d3c/dlib/cuda/cuda_dlib.cu#L1457-L1481

At least that's how `leaky_relu` handles the inplace case... But I am clearly missing something

Reply:
Right, but `computed_output` just isn't the same thing as `sub.get_output()`.  This PR changes one of the input from `sub.get_output()` to `computed_output`, so the output isn't going to be the same.

Reply:
Oh, I thought that was also handled automatically by the framework (when calling the `_inplace` variants).
So, I need to figure out how to update the gradient call now... I've been staring for an embarrassingly long time at both cuda mish implementations and leaky_relu implementations, and I am having a hard time figuring out why it doesn't work.
I get your point, that before I was acting on the input of the mish layer, and now I am acting on the ouput of the mish layer, so the gradients must obviously be different.

Reply:
It’s a matter of computing the gradient. The existing mish gradient function just doesn’t compute the gradient inplace since it requires knowing the old inputs, which will have been overwritten if the layer is an inplace layer. So you need a new function.

Reply:
Thank you for the explanation, and now I see that the `leaky_relu` gradient only depends on the `gradient_input`, but I find it a bit confusing that the method is called gradient_inplace, does that mean that the current inplace implementation in dlib is wrong?

Reply:
I mean this one:
https://github.com/davisking/dlib/blob/3162f93c5d303e593337ab8e40d79b385e025d3c/dlib/cuda/cuda_dlib.cu#L1457-L1461